keystone/common/api_filtering.py

```diff
@@ -0,0 +1,210 @@
+                default=['admin', 'service', 'monitoring', 'support'],
+                help='List of projects to filter by default.'),
+    cfg.ListOpt('roles_to_filter',
+                default=['admin', 'KeystoneAdmin',
+                         'KeystoneServiceAdmin', 'service',
+                         'heat_domain_admin', 'support'],
+                help='List of roles to filter by default.'),
+    cfg.ListOpt('explicit_users_to_filter',
+                default=['heat_domain_admin'],
+                help='List of users to filter by default.'),
+    cfg.ListOpt('admin_roles',
+                default=['cloud_admin', 'project_admin'],
+                help='List of admin roles by default.')]
+
+CONF = config.CONF
+CONF.register_group(opt_group)
+CONF.register_opts(opts, opt_group)
+
+
+def filter_projects(f):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+
+        res = f(*args, **kwargs)
+        is_admin_flag = is_admin(args[0], args[1])
+
+        if not is_admin_flag:
+            if kwargs is not None and kwargs.get('project_id') is not None:
+                if (res['project']['name']
+                        in CONF.api_filtering.projects_to_filter):
+                    msg = ('Project \'%s\' does not exist.'
+                           % kwargs.get('project_id'))
+                    LOG.error(_LE(msg))
+                    raise exception.NotFound(msg)
+            else:
+                refs = [i for i in res['projects']]
+                for i in refs:
+                    if (i['name']) in CONF.api_filtering.projects_to_filter:
+                        res['projects'].remove(i)
+        return res
+    return wrapper
+
+
+def filter_roles(f):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+
+        res = f(*args, **kwargs)
+        is_admin_flag = is_admin(args[0], args[1])
+
+        if not is_admin_flag:
+            if kwargs is not None and kwargs.get('role_id') is not None:
+                if res['role']['name'] in CONF.api_filtering.roles_to_filter:
+                    msg = ('Role \'%s\' does not exist.'
+                           % kwargs.get('role_id'))
+                    LOG.error(_LE(msg))
+                    raise exception.NotFound(msg)
+            else:
+                refs = [i for i in res['roles']]
+                for i in refs:
+                    if (i['name']) in CONF.api_filtering.roles_to_filter:
+                        res['roles'].remove(i)
+        return res
+    return wrapper
+
+
+def filter_users(f):
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+
+        project_details = {}
+        res = f(*args, **kwargs)
+
+        is_admin_flag = is_admin(args[0], args[1])
+
+        if not is_admin_flag:
+            projects = list_projects(args[0], args[1])
+
+            for project in projects['projects']:
+                if project['name'] in CONF.api_filtering.projects_to_filter:
+                    project_details[project['name']] = project['id']
```

Comment:
Seems odd to be creating a key named `name` but then populating it with the `id`.